subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/CallInterceptorsSet.java

```diff
@@ -223,21 +226,36 @@ public Object callOriginal() throws Throwable {
             }
         }
 
-        private InstrumentedGroovyCallsHelper.ThrowingCallable<Object> callableForOriginalCall(
-            CallStrategy callStrategy,
-            Object receiver,
-            @Nullable Object[] args
-        ) {
-            InstrumentedGroovyCallsHelper.ThrowingCallable<Object> result =
-                callStrategy == CallStrategy.CALL_CURRENT ? () -> super.callCurrent((GroovyObject) receiver, args) :
-                    callStrategy == CallStrategy.CALL_GROOVY_OBJECT_GET_PROPERTY ? () -> super.callGroovyObjectGetProperty(receiver) :
-                        null;
-
-            if (result == null) {
-                throw new IllegalArgumentException("unexpected original call strategy " + callStrategy);
+        @SuppressWarnings("DanglingJavadoc")
+        private Object invokeDefaultGroovyCallSiteImplementation(Object receiver, @Nullable Object[] args, CallStrategy callStrategy) throws Throwable {
+            if (callStrategy != CallStrategy.CALL_CURRENT && callStrategy != CallStrategy.CALL_GROOVY_OBJECT_GET_PROPERTY) {
+                throw new IllegalArgumentException();
             }
 
-            return result;
+            if (groovyDefaultCallSite != null) {
+                assert groovyDefaultCallSite != this;
+                return callStrategy == CallStrategy.CALL_CURRENT ? groovyDefaultCallSite.callCurrent((GroovyObject) receiver, args) : groovyDefaultCallSite.callGroovyObjectGetProperty(receiver);
+            } else {
+                try {
+                    return callStrategy == CallStrategy.CALL_CURRENT ? super.callCurrent((GroovyObject) receiver, args) : super.callGroovyObjectGetProperty(receiver);
+                } finally {
+                    /**
+                     * The default Groovy implementation replaces the entry in the call site array with what it creates based on the call kind. <p>
+                     * For example, see this code path:
+                     * {@link AbstractCallSite#callCurrent(GroovyObject, Object[])}
+                     * -> {@link org.codehaus.groovy.runtime.callsite.CallSiteArray#defaultCallCurrent}
+                     * -> {@link org.codehaus.groovy.runtime.callsite.CallSiteArray#createCallCurrentSite(CallSite, GroovyObject, Object[], Class))
+                     * -> {@link org.codehaus.groovy.runtime.callsite.CallSiteArray#replaceCallSite(CallSite, CallSite)} <p>
+                     * Because of it doing so, our decorated call site is removed from the dynamic invocation code path, and we lose the ability to track dynamic call entry points. <p>
```

Comment:
❓ Will the Groovy runtime replace the call site every time or just the first?

Reply:
It's in the implementation of `AbstractCallSite` (which does not actually know how to call anything) with a more concrete (and optimized) subtype, based on what method is called on `AbstractCallSite`. If we go through `super.*` once again, it will do the same. But we can cache the more concrete call site created in the first invocation, and calls to it won't modify the call site array.